Cython/Utility/Profile.c

```diff
@@ -33,8 +33,229 @@
+  // We therefore use a different name for the different types.
+  #define __PYX_MONITORING_ABI_SUFFIX  "_mon"
+#else
+  #define __PYX_MONITORING_ABI_SUFFIX
+#endif
+
 #if CYTHON_PROFILE || CYTHON_TRACE
 
+// instruction offset == (line << 9) | column
+#define __Pyx_offset_to_line(offset)  ((offset) >> 9)
+
+#if CYTHON_USE_SYS_MONITORING
+// Use the Py3.13 monitoring C-API: https://github.com/python/cpython/issues/111997
+
+  // TODO: make list of event types specific to functions
+
+  typedef enum {
+      __Pyx_Monitoring_PY_START = 0,
+      __Pyx_Monitoring_PY_RETURN,
+      __Pyx_Monitoring_PY_UNWIND,
+      //__Pyx_Monitoring_CALL,
+      __Pyx_Monitoring_LINE,
+      __Pyx_Monitoring_RAISE,
+      __Pyx_Monitoring_RERAISE,
+      __Pyx_Monitoring_EXCEPTION_HANDLED,
+      __Pyx_Monitoring_PY_RESUME,
+      __Pyx_Monitoring_PY_YIELD,
+      __Pyx_Monitoring_STOP_ITERATION,
+  } __Pyx_Monitoring_Event_Index;
+
+  static const unsigned char __Pyx_MonitoringEventTypes[] = {
+      PY_MONITORING_EVENT_PY_START,
+      PY_MONITORING_EVENT_PY_RETURN,
+      PY_MONITORING_EVENT_PY_UNWIND,
+      //PY_MONITORING_EVENT_CALL,
+      PY_MONITORING_EVENT_LINE,
+      PY_MONITORING_EVENT_RAISE,
+      PY_MONITORING_EVENT_RERAISE,
+      PY_MONITORING_EVENT_EXCEPTION_HANDLED,
+      // generator specific:
+      PY_MONITORING_EVENT_PY_RESUME,
+      PY_MONITORING_EVENT_PY_YIELD,
+      PY_MONITORING_EVENT_STOP_ITERATION,
+  };
+
+  typedef uint64_t __pyx_monitoring_version_type;
+
+  #define __Pyx_MonitoringEventTypes_CyFunc_count (sizeof(__Pyx_MonitoringEventTypes) - 3)
+  #define __Pyx_MonitoringEventTypes_CyGen_count (sizeof(__Pyx_MonitoringEventTypes))
+
+  #define __Pyx_TraceDeclarationsFunc \
+      PyObject *$frame_code_cname = NULL; \
+      PyMonitoringState $monitoring_states_cname[__Pyx_MonitoringEventTypes_CyFunc_count]; \
+      int __pyx_exception_already_reported = 0;
+
+  #define __Pyx_TraceDeclarationsGen \
+      PyObject *$frame_code_cname = $generator_cname->gi_code; \
+      PyMonitoringState* $monitoring_states_cname = $generator_cname->$monitoring_states_cname; \
+      __pyx_monitoring_version_type $monitoring_version_cname = $generator_cname->$monitoring_version_cname; \
+      int __pyx_exception_already_reported = 0;
+
+  #define __Pyx_IsTracing(event_id)  (($monitoring_states_cname[event_id]).active)
+  #define __Pyx_TraceFrameInit(codeobj) \
+      if (codeobj) $frame_code_cname = codeobj;
+
+  CYTHON_UNUSED static PyCodeObject *__Pyx_createFrameCodeObject(const char *funcname, const char *srcfile, int firstlineno); /*proto*/
+  CYTHON_UNUSED static int __Pyx__TraceStartFunc(PyMonitoringState *state_array, PyObject *code_obj, int offset, int skip_event); /*proto*/
+  CYTHON_UNUSED static int __Pyx__TraceStartGen(PyMonitoringState *state_array, __pyx_monitoring_version_type *version, PyObject *code_obj, int offset); /*proto*/
```

Comment:
There is also macro There is macro with name `__Pyx_TraceStartFunc`. The name is confusing because the two names (macro and static function) are the same except one extra `_`. 

The same note holds also for rest of function defined here.

Reply:
With that said - there may be a general point here that the macros very much macro-like macros (rather than function-like macros) given the use of `error_goto`. So maybe should be named as macros. I know these are pre-existing though.

Reply:
They could arguably have more macro-ish names, but just some of them make use of the error-goto. Even those could be replaced by a function call + "if -1 goto error". The reason for macros is mostly the inlined is-tracing check and the nogil handling (which, arguably, is a bit of laziness because we could also use different functions for the no-gil and have-gil cases).

I'll stick with the current names, they look at least somewhat consistent.